src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,750 @@
+                        );
+                    }
+                    None
+                }
+            }
+        });
+        match ns {
+            Some(ns) => (ns.db, ns.coll),
+            None => (
+                self.default_namespace.db.clone(),
+                self.default_namespace.coll.clone(),
+            ),
+        }
+    }
+
+    fn append(
+        &self,
+        insert_builder: &mut HashMap<MongodbNamespace, InsertCommandBuilder>,
+        row: RowRef<'_>,
+    ) -> Result<()> {
+        let document = self.row_encoder.encode(row)?;
+        let ns = self.extract_namespace_from_row_ref(row);
+        let coll = ns.1.clone();
+        match insert_builder.entry(ns) {
+            Entry::Occupied(mut entry) => entry.get_mut().append(document),
+            Entry::Vacant(entry) => {
+                let mut builder = InsertCommandBuilder::new(coll, self.max_entries);
+                builder.append(document);
+                entry.insert(builder);
+            }
+        }
+        Ok(())
+    }
+
+    fn upsert(
+        &self,
+        upsert_builder: &mut HashMap<MongodbNamespace, UpsertCommandBuilder>,
+        op: Op,
+        row: RowRef<'_>,
+    ) -> Result<()> {
+        let mut document = self.row_encoder.encode(row)?;
+        let ns = self.extract_namespace_from_row_ref(row);
+        let coll = ns.1.clone();
+
+        let pk = self.row_encoder.construct_pk(row);
+
+        // Specify the primary key (_id) for the MongoDB collection if the user does not provide one.
+        if self.pk_indices.len() > 1
+            || self.schema.fields[self.pk_indices[0]].name != MONGODB_PK_NAME
+        {
+            // compound pk should not have a field named `_id`
+            document.insert(MONGODB_PK_NAME, pk.clone());
+        }
+
+        let pk = doc! {MONGODB_PK_NAME: pk};
+        match op {
+            Op::Insert | Op::UpdateInsert => match upsert_builder.entry(ns) {
+                Entry::Occupied(mut entry) => entry.get_mut().add_upsert(pk, document),
+                Entry::Vacant(entry) => {
+                    let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                    builder.add_upsert(pk, document)?;
+                    entry.insert(builder);
+                    Ok(())
+                }
+            },
+            Op::UpdateDelete => Ok(()),
+            Op::Delete => match upsert_builder.entry(ns) {
+                Entry::Occupied(mut entry) => entry.get_mut().add_delete(pk),
+                Entry::Vacant(entry) => {
+                    let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                    builder.add_delete(pk)?;
+                    entry.insert(builder);
+                    Ok(())
+                }
+            },
+        }
+    }
+
+    async fn write(&mut self, op: Op, row: RowRef<'_>) -> Result<()> {
+        let mut command_builder = self.command_builder.take();
```

Comment:
It looks wired to always to take `command_builder` out here and then set it back at the end.

I understand that you implement it in this way because otherwise `command_builder` will be bundled with `&mut self` and then you can't call `self.append`. 

A perfect solution will be to separate the mutable and immutable fields in the struct into different structs, so that we can both carry the used immutable fields via `self` and also modify the mutable fields, but this may take some efforts to refactor the code. A quick workaround is that we don't reuse the builder we obtained here by the enum match, but rematch the enum in `append` and `upsert`. Code will be like
```
async fn write(&mut self, ...) {
    if self.command_builder.is_append_only() {
        self.append(...).await
    } else {
        self.upsert(...).await
    }
}

async fn append(&mut self, ...) {
    let insert_builder = must_match!(&mut self.command_builder, CommandBuilder::AppendOnly(builder) => builder);
    ...
}

async fn upsert(&mut self, ...) {
    let upsert_builder = must_match!(&mut self.command_builder, CommandBuilder::Upsert(builder) => builder);
    ...
}
```

Besides, it may incur unnecessary cost to do enum match for each single row. We can instead only do enum match for each single stream chunk like the following
```
async fn write_batch(&mut self, chunk: StreamChunk) {
    if self.command_builder.is_append_only() {
        self.append(chunk).await
    } else {
        self.upsert(chunk).await
    }
}

async fn append(&mut self, chunk: StreamChunk) {
    let insert_builder = must_match!(&mut self.command_builder, CommandBuilder::AppendOnly(builder) => builder);
    for (op, row) in chunk.rows() {
        ...
    }
}

async fn upsert(&mut self, chunk: StreamChunk) {
    let upsert_builder = must_match!(&mut self.command_builder, CommandBuilder::Upsert(builder) => builder);
    for (op, row) in chunk.rows() {
        ...
    }
}
```